api/handlers.go

```diff
@@ -212,12 +240,50 @@ func (api *API) userGET(w http.ResponseWriter, req *http.Request, _ httprouter.P
 
 // userLimitsGET returns the speed limits which apply to this user.
 func (api *API) userLimitsGET(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
-	u := api.userFromRequest(req)
-	if u == nil || u.QuotaExceeded {
+	t, err := tokenFromRequest(req)
+	if err != nil {
 		api.WriteJSON(w, database.UserLimits[database.TierAnonymous])
 		return
 	}
-	api.WriteJSON(w, database.UserLimits[u.Tier])
+	tk, err := jwt.ValidateToken(t)
+	if err != nil {
+		api.WriteJSON(w, database.UserLimits[database.TierAnonymous])
+		return
+	}
+	s, exists := tk.Get("sub")
+	if !exists {
+		api.staticLogger.Tracef("Token without a sub.")
+		api.WriteJSON(w, database.TierAnonymous)
+		return
+	}
+	sub := s.(string)
+	// If the user is not cached, or they were cached too long ago we'll fetch
+	// their data from the DB.
+	userLimitCacheMu.Lock()
+	defer userLimitCacheMu.Unlock()
+	if c, exists := userLimitCache[sub]; !exists || c.LastUpdate.Sub(time.Now()) > userLimitCacheTTL {
+		// we don't want to hold the lock during a DB operation
+		userLimitCacheMu.Unlock()
+		u, err := api.staticDB.UserBySub(req.Context(), sub, false)
+		userLimitCacheMu.Lock()
+		if err != nil {
+			api.staticLogger.Debugf("Failed to fetch user from DB for sub '%s'. Error: %s", sub, err.Error())
+			api.WriteJSON(w, database.TierAnonymous)
+			return
+		}
+		if u == nil || u.QuotaExceeded {
+			userLimitCache[sub] = userLimitCacheEntry{
+				Tier:       database.TierAnonymous,
+				LastUpdate: time.Now().UTC(),
```

Comment:
My spider-sense tells me that there is some danger here.

I think you need to make sure that you set `LastUpdate` in a way that the cache invalidation aligns with a potential quota refill at the end of the month. Otherwise the user might end up getting the Anonymous tier for 1 hour after they should already be back to the tier they pay for.